types/index.d.ts

```diff
@@ -0,0 +1,100 @@
+// TypeScript Version: 2.4
+// Project: Koa Pageable
+// Definitions by: Wolf Rourke <wolf@panderalabs.com>
+import * as Koa from 'koa';
+
+export enum Direction {
+  asc = 'asc',
+  desc = 'desc',
+}
+
+export class Order {
+  static DEFAULT_DIRECTION: string;
+
+  direction: Direction;
+  property: string;
+  constructor(property: string, direction?: Direction | string);
+}
+
+export class Sort {
+  orders: Order[];
+  constructor(orders: Order[]);
+  [Symbol.iterator]: any;
+  forEach(iteratee: (property: string, direction: Direction) => any): void;
+  toJSON(): Order[];
+}
+
+export class Pageable {
+  indexed: boolean;
+  page: number;
+  size: number;
+  sort: Sort | null;
+  constructor(
+    pageNumber: number,
+    pageSize: number,
+    indexed: boolean,
+    sort: string | string[] | Sort | null,
+  );
+}
+
+export class Page<T> extends BasePage {
+  content: T[];
+  constructor(content: T[], totalElements: number, pageable: Pageable);
+  map<R>(iteratee: (_: T) => R): Page<R>;
+}
+
+export class BasePage {
+  first: boolean;
+  last: boolean;
+  number: number;
+  numberOfElements: number;
+  size: number;
+  sort: any;
+  totalElements: number;
+  totalPages: number;
+  constructor(totalElements: number, pageable: Pageable);
+}
+
+export class IndexedPage<I extends string, T> extends BasePage {
```

Comment:
This requires our IDs to be strings. A generic (I) is not allowed when we're defining the index of an object (line 63 below) - it must be either a string or a number and cannot be a union type.